### PR TITLE
UNI-528 Fixing param finder bug

### PR DESCRIPTION
**Pull request by marionleborgne**
Fixed issue with param finder collapsing milliseconds values into the same second, leading to an incorrect value for `medianSampleInterval`.

cc @lscheinkman @cbaranski @ywcui1990

### CODE REVIEW

File: unicorn/py/unicorn_backend/param_finder.py [hunk 1]
```diff
@@ -228,37 +228,40 @@ def findParameters(samples):
     }
     return outputInfo
 
-  timestamps = numpy.array(timestamps, dtype="datetime64[s]")
+  # make sure that timestamps are parsed in ms
+  timestampsInMs = numpy.array(timestamps, dtype="datetime64[ms]")
```
Comment (marionleborgne) [inline unicorn/py/unicorn_backend/param_finder.py]: This is where the source of the problem was

**Comment by marionleborgne** [comment]
You can test it by starting the app and uploading this file: https://github.com/marionleborgne/numenta-apps/blob/a2669710de2c5bdb53c821c12eb203ddd4fa35cf/unicorn/tests/js/fixtures/unix-timestamp-milliseconds-as-seconds.csv

**Comment by marionleborgne** [comment]
before testing make sure to run:
```
cd unicorn/app
npm run install:backend
```

To run the unit, integration and NAB compatibility tests:
```
cd unicorn/
npm run test:backend
```

**Comment by marionleborgne** [comment]
On master with the same file [1] you get:

screen shot 2016-06-14 at 2 27 04 pm

But with this new fix the param finder does not fail anymore 😄 and you get results:
screen shot 2016-06-14 at 2 19 43 pm
 
(BTW this is Harry's data which originally has unix timestamps - milliseconds converted to floating point seconds) 

💃

[1] https://github.com/marionleborgne/numenta-apps/blob/a2669710de2c5bdb53c821c12eb203ddd4fa35cf/unicorn/tests/js/fixtures/unix-timestamp-milliseconds-as-seconds.csv

**Comment by lscheinkman** [comment]
@marionleborgne It is not working for me. Here is the error:
```
22:45:31.697Z ERROR htmstudio:  (metricId=57f317bec87020b2!4f6b07a1045098b9)
err: {
"errorText": "year is out of range", 
"diagnosticInfo": "Traceback (most recent call last):
  File \"build/bdist.macosx-10.5-x86_64/egg/unicorn_backend/param_finder_runner.py\", line 201, in main
  File \"build/bdist.macosx-10.5-x86_64/egg/unicorn_backend/param_finder_runner.py\", line 182, in _readCSVFile
  File \"build/bdist.macosx-10.5-x86_64/egg/unicorn_backend/utils/date_time_utils.py\", line 87, in parseDatetime
    return datetime.utcfromtimestamp(timestampFloat)\nValueError: year is out of range"
}
    
    --
    inputOpt: {
      "csv": "~/Library/Application Support/HTM Studio/files/unix- 400 rows no blank row.csv",
      "rowOffset": 1,
      "timestampIndex": 0,
      "valueIndex": 3,
      "datetimeFormat": "#T"
    }
22:45:31.713Z ERROR htmstudio:  (metricId=57f317bec87020b2!4f6b07a1045098b9, event=close, code=1)
    inputOpt: {
      "csv": "~/Library/Application Support/HTM Studio/files/unix- 400 rows no blank row.csv",
      "rowOffset": 1,
      "timestampIndex": 0,
      "valueIndex": 3,
      "datetimeFormat": "#T"
    }

```

**Comment by lscheinkman** [comment]
@marionleborgne It worked with `/unicorn/tests/js/fixtures/unix-timestamp-milliseconds-as-seconds.csv`

**Comment by lscheinkman** [comment]
However it does not work with `unix- 400 rows no blank row.csv` or the other files attached to the JIRA

**Comment by lscheinkman** [comment]
got this error on `machine_data_2` file
```
    Error: {"errorText": "list index out of range", "diagnosticInfo": 
"Traceback (most recent call last):
  File \"build/bdist.macosx-10.5-x86_64/egg/unicorn_backend/model_runner_2.py\", line 414, in main
  File \"build/bdist.macosx-10.5-x86_64/egg/unicorn_backend/model_runner_2.py\", line 324, in run
IndexError: list index out of range
"}
```

**Comment by andrewmalta13** [comment]
in reply to https://github.com/numenta/numenta-apps/pull/853#issuecomment-226039280

this error is coming form date_time_utils.py UNI-545 is related.

**Comment by marionleborgne** [comment]
@lscheinkman the file in the JIRA has been open with excel and the auto-formatting in excel removed all the decimals. Hence all the timestamps are the same (they are all equal to `1.45E+12`). See below:

```

time,Gx,Gy,Gz
1.45E+12,279,-391,117
1.45E+12,-85,-153,168
1.45E+12,-242,76,186
1.45E+12,-354,-305,122
1.45E+12,-554,-213,178
1.45E+12,-657,-3,253
1.45E+12,-764,106,255
1.45E+12,-723,491,336
1.45E+12,-667,358,364
1.45E+12,-877,240,316
1.45E+12,-740,661,414
1.45E+12,279,-391,117
1.45E+12,-85,-153,168
1.45E+12,-242,76,186
```


There is nothing that the param finder can find in this file. I have the original file, and that's that one that I attached in the PR (https://github.com/marionleborgne/numenta-apps/blob/a2669710de2c5bdb53c821c12eb203ddd4fa35cf/unicorn/tests/js/fixtures/unix-timestamp-milliseconds-as-seconds.csv) 

If you try it with this one, it should work.

**Comment by marionleborgne** [comment]
In reply to https://github.com/numenta/numenta-apps/pull/853#issuecomment-226039280. I agree with @andrewmalta13 : once UNI-545 is merged, the new conversion in `date_time_utils.py` will make this problem go away

**Comment by lscheinkman** [comment]
`withTZ - 400 rows.csv` fails with "divide by zero":
```
    err: app/node_modules/portable_python_darwin/lib/python2.7/site-packages/unicorn_backend-0.0.0-py2.7.egg/unicorn_backend/param_finder.py:366: RuntimeWarning: divide by zero encountered in divide
    
    --
    inputOpt: {
      "csv": "~/Library/Application Support/HTM Studio/files/withTZ - 400 rows.csv",
      "rowOffset": 1,
      "timestampIndex": 0,
      "valueIndex": 1,
      "datetimeFormat": "%Y-%m-%dT%H:%M:%S.%f%z"
    }

```

**Comment by lscheinkman** [comment]
The one in the test worked (https://github.com/numenta/numenta-apps/pull/853#issuecomment-226040485) however I was validating the files in the JIRA first

**Comment by lscheinkman** [comment]
I will try to get original `unix- 400 rows no blank row.csv` file from celeste

**Comment by marionleborgne** [comment]
@lscheinkman Thanks for pointing out that it does not work with `withTZ - 400 rows.csv`. I'll add a fix to make sure it's never divided by zero

**Comment by marionleborgne** [comment]
@lscheinkman I added a fix for the `divided by zero error`. Note that you need to run the command below again for this change to take effect
```
cd unicorn/app/
npm install:backend 
```


Also note that the file `withTZ - 400 rows.csv` is kind of invalid. It has only 2 different timestamps: `2016-04-21T21:15:00.000+1700` and `2016-04-21T21:30:00.000+1700` which is why the median sampling interval is zero. See below [1]. To help @cbaranski , I generated another file in the same flavor but with non repeating timestamps: [withTZ-400-new.csv.zip](https://github.com/numenta/numenta-apps/files/315432/withTZ-400-new.csv.zip)


[1] content of `withTZ - 400 rows.csv`
```
_time,server_latency
2016-04-21T21:15:00.000+1700,16791
2016-04-21T21:30:00.000+1700,11351
2016-04-21T21:15:00.000+1700,16791
2016-04-21T21:30:00.000+1700,11351
2016-04-21T21:15:00.000+1700,11351
2016-04-21T21:30:00.000+1700,11351
2016-04-21T21:15:00.000+1700,11351
2016-04-21T21:30:00.000+1700,11351
2016-04-21T21:15:00.000+1700,11351
```

**Comment by marionleborgne** [comment]
@lscheinkman ready for review again

**Comment by lscheinkman** [comment]
The new file `withTZ-400-new.csv.zip` failed with unsupported timestamp but it looks like ['ISO 8601 NO T'](http://numenta.com/htm-studio/date-time-formats.html) => `YYYY-MM-DD HH:mm:ssZ` 
screen shot 2016-06-15 at 9 05 24 am

**Comment by lscheinkman** [comment]
@marionleborgne `machine_data_2 - 44K rows duped timestamp.csv` still fails with `list index out of range`:
```
    Error: {"errorText": "list index out of range", "diagnosticInfo": "
Traceback (most recent call last):
  File \"build/bdist.macosx-10.5-x86_64/egg/unicorn_backend/model_runner_2.py\", line 414, in main
  File \"build/bdist.macosx-10.5-x86_64/egg/unicorn_backend/model_runner_2.py\", line 324, in run
IndexError: list index out of range
"}
        at Socket.child.stderr.on.error (ModelService.js:129:23)
        at emitOne (events.js:77:13)
        at Socket.emit (events.js:169:7)
        at readableAddChunk (_stream_readable.js:146:16)
        at Socket.Readable.push (_stream_readable.js:110:10)
        at Pipe.onread (net.js:523:20)
    --
    inputOpt: {
      "csv": "~/Library/Application Support/HTM Studio/files/machine_data_2 - 44K rows duped timestamp.csv",
      "rowOffset": 1,
      "timestampIndex": 0,
      "valueIndex": 1,
      "datetimeFormat": "%Y-%m-%d %H:%M:%S"
    }
```

**Comment by lscheinkman** [comment]
@marionleborgne The original `withTZ - 400 rows.csv` is not causing 'divide by zero` but it is stuck on an infinite loop somewhere in the python code, ignoring the errors. see atached
screen shot 2016-06-15 at 9 42 49 am

**Comment by marionleborgne** [comment]
@lscheinkman my last commit should fix the flying CPU issue

**Comment by marionleborgne** [comment]
@lscheinkman I am now working on the model_runner error (value out of range)

**Comment by marionleborgne** [comment]
@lscheinkman this is the file with the correct ISO format to test timezones 😅 

[withTZ-400-correct-format.csv.zip](https://github.com/numenta/numenta-apps/files/317271/withTZ-400-correct-format.csv.zip)

**Comment by marionleborgne** [comment]
@lscheinkman this is the file with the correct ISO format to test timezones 😅 

[withTZ-400-correct-format.csv.zip](https://github.com/numenta/numenta-apps/files/317272/withTZ-400-correct-format.csv.zip)

**Comment by marionleborgne** [comment]
@lscheinkman I found why the issue with the machine data was happening. It's because there are a couple of empty rows at the end of the file. I am going to make the code more robust against that.

**Pull request merged**